flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/OnlineKMeans.java

```diff
@@ -0,0 +1,409 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.clustering.kmeans;
```

Comment:
Since OnlineKMeans is a different algorithm from KMeans, should we move its files to a dedicated folder, e.g. `onlinekmeans`?

Reply:
I think keeping `OnlineKMeans` in `kmeans` package is more suitable. Because
- I hardly feel that `OnlineKMeans` and `KMeans` are not different algorithms. They are the same algorithm with different external condition. 
- In implementation, `OnlineKMeans` and `KMeans` have exclusively shared some infra codes, like `KMeansModelData`, `KMeansParams` and `findClosestCentroidId`. If they are put into different packages, I am worried that there will be high cohesion between `kmeans` and onlinekmeans` packages.

The reasons above are not definitive and I would still be glad to make the change if we have other more important reasons.

Reply:
Strictly speaking they are different algorithms due to the following reasons:
- When they are given the same bounded training data, KMeans will iterate the data multiple times and OnlineKMeans will iterate the data exactly once. Thus the algorithm is different. And the resulting model data and prediction accuracy will also be different.
- They have different behavior in terms of whether they support unbounded training data.

You are right that KMeans and OnlineKMeans implementations share some infra codes. But this seems more like an implementation detail that is more related to developer experience.

After thinking about this more, I don't have a strong opinion here and both approaches work for me. We can keep it as is and separate them into different packages in the future when needed.